New: Processor improvements

nzakas opened the pull request:
## Summary

This proposal provides a way to explicitly define which processor(s) to use for different files inside of configuration. It also allows the chaining of multiple processors to fully process a file.

## Related Issues

* https://github.com/eslint/eslint/issues/11035

platinumazure commented on designs/2018-processors-improvements/README.md:
```diff
@@ -0,0 +1,202 @@
+Definitions:
+
+* **Extension-named processor** is the original processor design, where a processor is named only by the file extension to which the process should apply.
+* **Named processor** is introduced in this design, where a processor is given a name just like you would name a rule. The processor is not tied to a specific file extension by default.
```
Do we want to state that a named processor must not begin with a dot (to avoid ambiguity with extension-named processor)? And/or do we want to impose a stronger restriction?

nzakas replied:
Yes, that's a good restriction to add in.

platinumazure commented on designs/2018-processors-improvements/README.md:
```diff
@@ -0,0 +1,202 @@
+
+### Add a `processors` Configuration Option
+
+In order to use a named processor, end-users would specify a `processors` option in their configuration. The option must be an array and must contain at least one string identifier. The string identifier must correspond to a named processor from a plugin and have the form *pluginName/processorName*, which would reference the `pluginName` processor in the `eslint-plugin-pluginName` plugin. For exa ...
```
```suggestion
In order to use a named processor, end-users would specify a `processors` option in their configuration. The option must be an array and must contain at least one string identifier. The string identifier must correspond to a named processor from a plugin and have the form *pluginName/processorName*, which would reference the `processorName` processor in the `eslint-plugin-pluginName` plugin. For example: 
```

platinumazure commented on designs/2018-processors-improvements/README.md:
````diff
@@ -0,0 +1,202 @@
+};
+```
+
+This configuration applies two named processors to all files matching `*.md`. First, the `markdown` named processor from `eslint-plugin-markdown` is applied. The result of that preprocessing is then passed to the `vue` named processor from `eslint-plugin-vue`. The result of that preprocessing is then passed to ESLint for linting. When ESLint has completed linting, the reverse sequence occurs: the ...
````
I think it might be worth noting that this requires the `vue` processor to be able to handle non-Vue/non-JS code gracefully. (For example, if the Markdown processor passed along all code blocks rather than just JS, then the Vue plugin needs to gracefully handle that.)

nzakas replied:
Yes, I'll update that.

platinumazure commented on designs/2018-processors-improvements/README.md:
```diff
@@ -0,0 +1,202 @@
+
+## Alternatives
+
+@mysticatea submitted an alternative design proposal: [rfs/#1](https://github.com/eslint/rfcs/pulls/1)
```
Do we want to have a compare/contrast note here (or in the other proposal)?

Maybe this raises a more general question: Where is the best place to discuss or debate alternative proposals for the same area/feature?

nzakas replied:
Yes, I will do that in the next revision.

mysticatea commented:
Thank you for this proposal :)

Sounds nice to me.
I think that plugins for other kinds than JavaScript may not work without processors. For example, `eslint-plugin-vue` uses the `postprocess` API to provide `<!-- eslint-disable -->`-like directive comment functionality. If the processor is missing, the users will see cryptic error messages. But plugins can contain configuration presets, so it will be no problem even if the end-users need additional setting.

But I have a concern.

- how do users configure processors to extract a specific kind of blocks?

Some kind of files can contain various kinds of code blocks. such as markdown.
`eslint-plugin-markdown` wants to extract only JS blocks. But in the original issue, users want to configure to extract other blocks than JS. This proposal doesn't look to include the way to configure it.

For example:

```js
// .eslintrc.js
module.exports = {
    overrides: [
        {
            files: "*.md",
            processors: ["markdown/markdown"], // want markdown processor to extract JS blocks
        },
        {
            files: "*.md",
            processors: ["markdown/markdown"], // want markdown processor to extract TS blocks
            parser: "typescript-eslint-parser", 
        },
        {
            files: "*.md",
            processors: ["markdown/markdown", "html/html"], // want markdown processor to extract HTML blocks
        },
        {
            files: "*.md",
            processors: ["markdown/markdown", "vue/vue"], // want markdown processor to extract Vue blocks
            parser: "vue-eslint-parser", 
        },
    ]
}
```

May the `eslint-plugin-markdown` author has to define processors by the number of block kinds?

mysticatea commented:
After I re-looked my comment, I'm not sure that setting works well because the `processors` conflicts. But I think, that is the matter we want to solve.

```js
// .eslintrc.js
module.exports = {
    overrides: [
        {
            files: "*.md",
            processors: ["markdown/markdown"], // want markdown processor to extract JS blocks
        },
        {
            files: "*.md",
            processors: ["markdown/markdown"], // want markdown processor to extract TS blocks
            parser: "typescript-eslint-parser", 
        },
        {
            files: "*.md",
            processors: ["markdown/markdown", "html/html"], // want markdown processor to extract HTML blocks
        },
        {
            files: "*.md",
            processors: ["markdown/markdown", "vue/vue"], // want markdown processor to extract Vue blocks
            parser: "vue-eslint-parser", 
        },
    ]
}
```

mysticatea commented:
Ah, mmm, #1 also cannot solve that problem properly, because #1 will implement recursive calls in `Linter#verify`, but resolving `overrides` setting does in `CLIEngine`. I'm thinking...

nzakas commented:
@mysticatea 

> how do users configure processors to extract a specific kind of blocks?

It's not clear from the proposal, but this intentionally does not solve that problem because I don't believe that's what https://github.com/eslint/eslint/issues/11035 is asking for or needs. (I will update the proposal to make it this clear.)

So far, the only plausible use case would be Markdown with different embedded code fragments inside, maybe some that are JS and some that are TypeScript, etc. I think that this is a corner case to the larger issue of making processors end-user configurable. I would suggest we don't focus on solving that case right now.

That said, I don't believe that this proposal prevents a solution like #1 from being implemented at some point in the future. As you stated, the processing would have to move to outside of `Linter` into `CLIEngine` so that patterns could be match for the virtual filenames. If we find that we really do need this capability, then we can always go down that route later (and use `overrides` to match the virtual filenames to appropriate configs).

nzakas commented:
@ilyavolodin you are correct, this does not change how `--ext` works. I'll add a note to make that clear.

nzakas commented:
@mysticatea I'm sorry, I completely misunderstood our earlier discussion. You can pretty much ignore everything in my previous comment. Let me get in the latest updates and I'll revisit.

nzakas commented:
Okay, after looking this over again, I realize I completely missed the point that we need to be able to derive virtual filenames for extracted code blocks in this proposal. My apologies for creating any confusion.

I'm going to update this proposal to include @mysticatea's approach of providing a virtual filename for extracted code blocks as I think that's clearly the best way to solve that part of the problem. I think I can add that feature into this proposal without any significant changes, and if @mysticatea likes the updated proposal, I'd suggest we work together on a single proposal moving forward.

nzakas commented:
Okay, I've updated the proposal to take everyone's feedback into account. I added a bit about virtual filenames from @mysticatea's proposal and how the current implementation of processors would have to change. I *think* I now properly understand what we're going for, thanks for your patience and help.

mysticatea commented:
@nzakas Merging is a good idea. I have been thinking #1 should use `overrides[].processor` instead of processor API's extensions.

mysticatea commented on designs/2018-processors-improvements/README.md:
```diff
@@ -0,0 +1,271 @@
+module.exports = {
+    preprocess(text, filename) {
+        
+        const {vueText, jsText } = doSomething(text, filename);
```
How does the processor determine whether it should return `vueText` as well?

In general, ESLint can handle only JavaScript, so processor should return only JavaScript blocks.
For example, maybe `eslint-plugin-markdown` needs to have `markdown/js`, `markdown/js,html`, `markdown/js,ts`, `markdown/js,vue`, `markdown/js,html,ts`, `markdown/js,html,vue`, `markdown/js,ts,vue`, and `markdown/js,html,ts,vue` if there is no configuration way.

nzakas replied:
That's a good point. I was trying to make up a useful example but I think I confused the point. I will fix this.

mysticatea commented on designs/2018-processors-improvements/README.md:
```diff
@@ -0,0 +1,271 @@
+
+When a `preprocess()` method returns an object with a `filePath` property, `CLIEngine` will call `getConfigForFile()` on the `filePath` property to determine the correct configuration for the code block.
+
+When a `preprocess()` method returns only a string, `CLIEngine` will interpret that as a JavaScript file and call `getConfigForFile()` using the parent file's filename to determine the correction configuration for the code block.
```
I don't think that this process is needed because the configuration file will never change in this case.

nzakas replied:
The file doesn't change, but the `overrides` needs to be applied for each piece coming from a processor. That happens inside of `getConfigForFile()` right now, so calling `getConfigForFile()` for the parent file and for a code block with a different extension might return different configs.

nzakas replied:
Yes. I will make that clearer in the text.

mysticatea commented on designs/2018-processors-improvements/README.md:
```diff
@@ -0,0 +1,271 @@
+
+Currently, application of processors happens inside of the `Linter#verify()` and `Linter#verifyAndFix()` methods. In order to implement this proposal, application of processors must happen outside of `Linter` because configuration calculation happens outside of `Linter`.
+
+The application of processors must move into `CLIEngine#executeOnText()` and `CLIEngine#executeOnFiles()` to work alongside with the current configuration calculation. That also means `CLIEngine` becomes responsible for recombining and postprocessing results from multiple `Linter#verify()` calls.
```
I'm worried about online demos. If the processors logic is removed from `Linter`, the plugins which have processors cannot make their online demos. For example, https://mysticatea.github.io/vue-eslint-demo/ is using processors API.

mysticatea replied:
I hope to make a restriction that `preprocess` cannot change the directory that the file belongs instead. As a result, it doesn't need re-looking up configuration files by every preprocess, then it will avoid performance impact.

nzakas replied:
> If the processors logic is moved to CLIEngine, the plugins which have processors cannot use new processors API because of their online demos.

I'm not advocating moving the logic from `Linter` to `CLIEngine`. I'm proposing that we leave the `Linter` logic in place but have something in `CLIEngine` that is used with the CLI. So the CLI would never use the `Linter` logic but online demos would still be able to work the same as today. Does that make sense?

> I hope to make a restriction that preprocess cannot change the directory that the file belongs instead.

That's a good point! Should we make it so `preprocess` can only return a `filename` that disallows the `\` character?

mysticatea replied:
> Should we make it so preprocess can only return a filename that disallows the \ character?

Sounds good to me.

> I'm not advocating moving the logic from Linter to CLIEngine. I'm proposing that we leave the Linter logic in place but have something in CLIEngine that is used with the CLI. So the CLI would never use the Linter logic but online demos would still be able to work the same as today. Does that make sense?

Indeed, it will not break the existing online demos. But I think that it means we cannot use the new feature in online demos (it can be a blocker for migration). I had imaged that we can separate the resolving `overrides` logic from the finding config files logic then can move the former into `Linter`.

Current behavior:

1. `CLIEngine` finds the proper config and resolves `overrides`. (depends on `fs`)
1. `CLIEngine` picks up the proper processor from the config.
1. `Linter` applies the processor to code.
1. `Linter` verifies the processed code with the config.

New behavior that I'm imaging:

1. `CLIEngine` finds the proper config. (depends on `fs`)
1. `Linter` applies processors recursively with the config which contain `overrides` property.
1. `Linter` verifies the processed code with the config.

mysticatea replied:
OK. I got it that moving of the logic can be a breaking change.
I'm still thinking that the resolving of `overrides` should not depend on `fs`, but I can separate it from this proposal.

nzakas replied:
Sounds good. Thanks @mysticatea.

nzakas commented on designs/2018-processors-improvements/README.md:
```diff
@@ -0,0 +1,271 @@
+
+### Apply Multiple Processors
+
+In special cases, end-users may want to apply multiple processors to files. This can be accomplished by listing multiple named processors using the `processors` configuration option. For example:
```
Oh yes, you are correct. So I guess what we really need is a `processor` configuration key that is just a string because the rest of the logic will be determined by matching `overrides`.

nzakas commented on designs/2018-processors-improvements/README.md:
```diff
@@ -0,0 +1,271 @@
+
+This configuration uses the same `vue` named processor from `eslint-plugin-vue` for both HTML and Markdown files, but only after an appropriate named processor is used to extract the Vue code.
+
+**Note:** It would be update to the Vue processor to understand the difference between different types of code it may be passed.
```
Heh yeah, it should read "It would be up to the Vue processor". I'll fix that, thanks.

nzakas commented on designs/2018-processors-improvements/README.md:
```diff
@@ -0,0 +1,271 @@
+        // this entry is added internally for backwards compatibility
+        {
+            files: ["*.md"],
+            processors: ["markdown/markdown"]
```
Yes, good catch. I'll fix that.

nzakas commented on designs/2018-processors-improvements/README.md:
```diff
@@ -0,0 +1,271 @@
+
+## Changes to `CLIEngine` and `Linter`
+
+Currently, application of processors happens inside of the `Linter#verify()` and `Linter#verifyAndFix()` methods. In order to implement this proposal, application of processors must happen outside of `Linter` because configuration calculation happens outside of `Linter`.
```
Yes. As I mentioned in the Backwards Compatibility section, I'm not advocating removing any functionality from `Linter`. We can keep it there and just not use it directly from the CLI.

As a side note, I've been thinking that it's time to rethink the end-user facing browser API, as I don't think throwing everything into `Linter` is sustainable for much longer. I have some ideas about that and will probably do an RFC after the holidays.

nzakas commented on designs/2018-processors-improvements/README.md:
```diff
@@ -0,0 +1,271 @@
+
+The `preprocess()` method inside of a processor can now return an array containing either strings (the current behavior) or an object with two properties:
+
+* `filePath` - (string) a virtual file path to map the code block to
```
Yeah, @mysticatea mentioned that above as well. I think we probably want to make this into `filename` instead of `filePath` and probably forbid slash characters altogether. Just waiting to see what @mysticatea thinks about that.

nzakas commented:
I've incorporated the latest round of feedback including:

* Change from `processors` to `processor`, as multiple processors will now be applied by matching the file extension of code blocks.
* Clarified that `Linter` functionality will not be removed, only worked around in the CLI.
* Fixed typos and clarified some points.

Outstanding issues:

* It seems like we don't want to be able to specify a full file path for code blocks. The question is, what do we want in its place? Should we only allow processors to specify extensions? Should we allow processors to specify full filenames and disallow `\` characters?

btmills commented on designs/2018-processors-improvements/README.md:
```diff
@@ -0,0 +1,226 @@
+
+This processor returns two code blocks containing JavaScript code. Each code block is given a virtual filename ending with `.js`.
+
+When a `preprocess()` method returns an object with a `filePath` property, `CLIEngine` will call `getConfigForFile()` on the `filePath` property to determine the correct configuration for the code block, which includes whether another processor should be run on the code block (matched by file extension).
```
At least in the case of the Markdown processor, since it can’t know which extensions have processors configured, I understand it will want to return every code block it encounters? For example, if `original.md` contains `js`, `html`, and `vue` code blocks, it would return `original.md.js`, `original.md.html`, and `original.md.vue`. Do we then need to run those through the `--ext` filter as well? Perhaps I invoke with `--ext .html`, so I’d expect the `html` code block to be linted, but not the `vue` code block.

mysticatea replied:
At initial, I had considered the same thing: we should filter extracted blocks with the `--ext` or glob patterns. However, I think currently, if a specific file specified (E.g., `eslint docs/README.md`), I'm not sure what the proper behavior is. I will be confused if ESLint changes the target code blocks by globs. So I want options for processors in config files.

btmills replied:
I think that makes the most sense. The Markdown processor would return every fenced code block, and the HTML processor would return every `<script>` tag, and ESLint core would be responsible for deciding which of those get linted.

How that interacts with `--ext` I’m not sure. I suppose you could argue that if I configure an `override` for `*.ts`, then I shouldn’t also have to specify `--ext .ts`. When plugins provided their own extensions, the `--ext` flag was necessary, but flipping the relationship around so that the extensions are configured in `overrides` might make it redundant.

Is appending an extension to the existing filename the best way? I think so, but I wanted to point out that it’s trivial for the Markdown processor to pull the tag from the fenced code block, but an HTML processor would need to map from the `<script>`’s MIME type to the appropriate extension.

mysticatea replied:
I don't oppose the way.

My small concern is, some users use glob patterns rather than `--ext` option to specify other files than JS (E.g., `src/**/*.{js,ts,tsx}`). But we can say that use `--ext` option to specify embedded codes other than JS.

nzakas replied:
Okay, it seems like we are in agreement that:

1. Processors can return all file types it thinks ESLint can handle.
2. ESLint will filter out any file types that aren't specified by `--ext` to avoid errors.
3. Users must use `--ext` even if they have `overrides` in their config; `overrides` is not a replacement for `--ext` (in the context of this proposal).

@btmills I think appending the file extension makes sense. My hunch is that most processors will only be returning `.js` files any way. As @mysticatea has pointed out a couple of times, we are really talking about an edge case with non-`.js` code blocks being returned from processors.

nzakas commented:
I've incorporated the latest round of feedback including:

* Change `filePath` to `filename` for code blocks.
* `filename` in code blocks cannot contain slash characters
* Clarified how we will treat code blocks without a `filename` specified
* Fixed typos.

Outstanding issues:

* How does this proposal interact with the `--ext` flag?
* Where should the processing take place in order to expose this functionality to browsers?

mysticatea commented:
How does this proposal handle autofix?
If some of the processors don't support autofix, broken `fix` properties will mix in the result. #1 mentioned it and removing the broken `fix` properties.

nzakas commented:
@mysticatea oh, I guess I thought autofixing would "just work." :) Let me read back over that part of your proposal and see what I missed.

mysticatea commented on designs/2018-processors-improvements/README.md:
```diff
@@ -0,0 +1,241 @@
+
+When a `preprocess()` method returns an object with a `filename` property, `CLIEngine` will call `getConfigForFile()` on the `filename` property to determine the correct configuration for the code block, which includes whether another processor should be run on the code block (matched by file extension).
+
+When a `preprocess()` method returns only a string, `CLIEngine` will interpret that as a JavaScript file and call `getConfigForFile()` using the parent file's filename appended with `".js"` (in case the parent filename does not end with `".js"`) to determine the correction configuration for the code block using this algorithm:
```
I found a new concern:

Some `eslint-plugin-vue` users have their setting on `overrides` section:

```yml
#...
overrides:
  - files: "*.vue"
    parser: vue-eslint-parser
    plugins:
      - vue
    rules:
      vue/foo: error
      vue/bar: error
#...
```

And the plugin has an extension-named processor.

In this case, are the `vue/*` rules will get ignored because the code will get `.js` extension?
It looks a breaking change.

nzakas replied:
Hmmm, I think you are right. This would break with the current proposal. What does `vue-eslint-parser` do?

mysticatea replied:
`vue-eslint-parser` parses `*.vue` files as HTML and generates [Vue.js Template AST](https://github.com/mysticatea/vue-eslint-parser/blob/59d52edb9d01989713f06c82618ce5c2b74a5646/docs/ast.md) from the first `<template>` element. The parser provides `parserServices` to access to the template AST. Also, it generates ESLint AST from the first `<script>` element and returns it.

On the other hand, the processor of `eslint-plugin-vue` provides the feature of `<!-- eslint-disable -->`-like directive comments. That `postprocess` method filters messages it should ignore. In fact, that `preprocess` method does nothing.

nzakas replied:
Ah interesting. Thanks for explaining.

So is there a way we can do a non-breaking change to allow this behavior? (I don't think so, but I'm not 100% sure?)

If not, how would you be able to use `vue-eslint-parser` with this proposal?

mysticatea replied:
I think that we can keep the current behavior as is:

- it keeps the filename as is if `preprocess` returned a string.
- because the filename was not changed, it doesn't re-resolve the configuration and lints the code.

nzakas replied:
Okay, just to be 100% sure, you're saying we can keep what's in the current proposal and it will not be a breaking change?

nzakas replied:
@mysticatea can you answer my last question please?

mysticatea replied:
I'm sorry for overlooking.

I meant that we can keep the behavior when a processor's preprocess returned a string, as same as the current behavior. I.e., if the preprocess returned a string without a filename, ESLint doesn't change the filename that ESLint is verifying and call `Linter#verify`.

nzakas replied:
Oh I see. And that will make the `vue-eslint-parser` use case work correctly? So this is NOT a breaking change?

mysticatea replied:
> And that will make the vue-eslint-parser use case work correctly? So this is NOT a breaking change?

Yes.

> an infinite loop

We can generalize it a bit. This will happen when a processor returned the same filename as parent's. My proposal (using the parent filename if a processor omitted a code block's filename) matches the condition.

My original proposal had a few lines to avoid infinity looping: https://github.com/eslint/rfcs/pull/1/files#diff-ae4914c864eeee75706376579948dcd1R152

nzakas commented on designs/2018-processors-improvements/README.md:
````diff
@@ -0,0 +1,241 @@
+});
+```
+
+**Note:** Any virtual filename that does not have an extension already specified by `--ext` will be ignored and will not be linted or autofixed.
````
Another concern that I thought of last night: right now, the `filename` parameter of `preprocess()` is guaranteed to be unique each time it is called. With the current proposal, it's possible for `filename` to not be unique because we allow `preprocess()` to assign whatever filename it wants. 

I think we need to do one of these:

1. Force `preprocess()` to only assign a file extension and not a full filename. We would then make each code block have a name of `parentFilename + index + ext`.
2. Use the returned filename from `preprocess()` as just part of the full filename, so `parentFilename + index + preprocessFilename`.

(We could also insert `block` or `codeblock` into each filename to make it easier to use with `overrides`.)

nzakas replied:
@mysticatea thoughts on this?

mysticatea replied:
The `parentFilename + index + preprocessFilename` sounds good. We can mention the `preprocessFilename` can be only file extension in the document. If a code block has that name, `preprocess` can return the name.

For example:

- `{code: "...", filename: "js"}` → `parentFilename.md/0.js`
- `{code: "...", filename: "child.js"}` → `parentFilename.md/0.child.js`

If we use `path.sep` at between the parent filename and preprocess filename, it will not conflict with other existing files.

nzakas replied:
I'd like to stick with just using a filename, as I think it's a bit confusing to allow either the extension or the filename in the same field.

I'll update the proposal for this part.

nzakas replied:
I've updated the proposal, please take a look at this section.

nzakas commented:
@mysticatea thanks for explaining. Could autofix just be handled in `CLIEngine`? If any processor doesn't have `shouldAutofix: true` then we just don't do any fixes?

mysticatea commented:
> Could autofix just be handled in CLIEngine?

Yes.

> If any processor doesn't have shouldAutofix: true then we just don't do any fixes?

Yes.
But it may be hard if we use recursive calls because the lint result doesn't have the information. So my proposal is to remove `fix` property from the lint messages that the processor which doesn't support autofix returned. Such a `fix` property is broken and it has caused a problem with editor integrations. (https://github.com/Microsoft/vscode-eslint/issues/185#issuecomment-267570350)

nzakas commented:
@mysticatea just to make sure I understand, you're saying:

1. After linting is finished, we should check if the processor has `shouldAutoFix`, if it's not `true`, we should remove the `fix` key from each message.
2. If there is another processor to be run and `shouldAutoFix` is not `true`, remove any remaining `fix` keys in each message.

Is that correct?

My only concern is that removing `fix` changes the shape of the object, and that could have performance implications.

My initial thought was just that as we are doing preprocessing, if we discover any processor without `shouldAutoFix` equal to `true` that we call `Linter#verify()` instead of `Linter#verifyAndFix()` and avoid doing any of the fixing.

mysticatea commented:
Yes, it's correct.

> My only concern is that removing fix changes the shape of the object, and that could have performance implications.

I'm worried about it, too.

> My initial thought was just that as we are doing preprocessing, if we discover any processor without shouldAutoFix equal to true that we call Linter#verify() instead of Linter#verifyAndFix() and avoid doing any of the fixing.

We cannot use `Linter#verifyAndFix()` because processors have to be handled in inside of the method to run multipass autofixing. `CLIEngine` has to make the fixed code after it merged all results of code blocks by post-process.

nzakas commented:
> We cannot use Linter#verifyAndFix() because processors have to be handled in inside of the method to run multipass autofixing. CLIEngine has to make the fixed code after it merged all results of code blocks by post-process.

Oh, I see. Thanks for explaining. In that case, can we simplify this process by just saying that if any processor is missing `shouldAutofix: true`, we don't do any autofixing?

mysticatea commented:
> In that case, can we simplify this process by just saying that if any processor is missing shouldAutofix: true, we don't do any autofixing?

I think that removing `fix` properties is the most simple way. Because new logic is recursive but the lint result, i.e. the return value, doesn't have the flag of fixable or not. The guarantee that all `fix` properties are not broken will make the logic simple.

Or, we can add the flag to the return value of the recursive function.

mysticatea commented:
I'm wondering that we can add a flag to `Linter#verify` method in order to stop calling the `fix` function in the reported problems. Then we can stop making `fix` properties in lint results if processors don't support autofix. It will improve performance in that situation.

nzakas commented:
@mysticatea I really don't want to add yet another option to `Linter#verify()`. That API is already overly complicated and I don't think adding another option for this use case makes much sense.

Hehe, all of this is why we didn't allow autofixing for processors at first. :)

I think we could do something like:

```js
function applyProcessors(sourceCode) {
   // recursively apply processors
   return {
        shouldAutofix: (from),
        codeBlocks: [...]
    };
}
```

Such that each recursive call to `applyProcessors()` can return whether its processor has `shouldAutofix`. If it ends up that `shouldAutofix` is returned as `false` from the top-level call, then we can avoid autofixing altogether. This would mean that if any processor in the chain doesn't have `shouldAutofix: true` then no autofixing is done.

mysticatea commented:
@nzakas The processor is the pair of preprocess and postprocess, so it will be:

```js
function verify(filename, code) {
    //....(resolve config)....
    let shouldAutofix = true
    const messagesList = preprocess(code, filename).map(item => {
        if (typeof item === "string") {
            return linter.verify(code, config, { ...options, filePath: filename })
        }

        // Recursive call.
        const retv = verify(item.filename, item.code)

        // Take the flag.
        if (!retv.shouldAutofix) {
            shouldAutofix = false
        }
        return retv.messages
    })

    return {
        shouldAutofix,
        messages: postprocess(messagesList, filename)
    }
}
```

As a reference, stopping creating `fix`:

```js
function verify(filename, code) {
    //....(resolve config)....
    return postprocess(
        preprocess(code, filename).map(item => {
            if (typeof item === "string") {
                return linter.verify(code, config, {
                    ...options,
                    filePath: filename,
                    disableConstructFix: !supportsAutofix
                })
            }
            // Recursive call.
            return verify(item.filename, item.code)
        }),
        filename,
    )
}
```

Actually, the `disableConstructFix` flag is unnecessary if the `overrides` resolving logic exists in `Linter` because `Linter` knows when it should not construct the `fix` property.

nzakas commented:
Ah ok, I think I see what you're getting at. I (reluctantly) agree that adding the flag to `Linter#verify()` to disable creating of `fix` makes sense.

So that `verify()` function in your example would end up in `CLIEngine`?

mysticatea commented:
> So that verify() function in your example would end up in CLIEngine?

Yes :smiley:

nzakas commented:
Okay, I'll update the design within the next couple of days.

On Tue, Dec 4, 2018 at 2:23 PM Toru Nagashima <notifications@github.com>
wrote:

> So that verify() function in your example would end up in CLIEngine?
>
> Yes 😃
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/rfcs/pull/3#issuecomment-444282300>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACWkuDo1YDzKaMJ-qDiGHzhZH4QDt31ks5u1vXogaJpZM4Yrmbj>
> .
>


-- 

______________________________
Nicholas C. Zakas
@slicknet

Author, Principles of Object-Oriented JavaScript <http://amzn.to/29Pmfrm>
Author, Understanding ECMAScript 6 <http://amzn.to/29K1mIy>

nzakas commented:
I've updated the design to address the edge cases @mysticatea pointed out. I believe that this design is now ready for a final review.

btmills commented on designs/2018-processors-improvements/README.md:
```diff
@@ -0,0 +1,284 @@
+
+## Backwards Compatibility Analysis
+
+This proposal is 100% backwards compatible until we remove the old way of defining processors. Both named and extension-based processors can be defined in the same plugin, such as:
```
I've spent a while trying to figure this out, so please correct me if I've missed something. I think that this will require a major version bump when plugins start to support this API.

Currently, as well as after this change in the case of extension-named processors, code blocks inherit their parent file's filename. For example, `project/README.md` would contain blocks named `project/README.md`. In the Markdown plugin's readme right now, we [recommend configuring rules etc. using `*.md` in `overrides`](https://github.com/eslint/eslint-plugin-markdown#project-or-directory-wide-overrides-for-code-snippets).

When a processor is upgraded to support the named processor API, those code blocks would become `project/README.md/0.js`, and any rules configured for `*.md` files would no longer be applied to the code blocks inside of them.

Individually, the `eslint` adding support for this API won't break the existing (now called extension-named) processors, and `eslint-plugin-markdown` can start exporting both extension-named and named processors without breaking versions of ESLint that don't support the new API. But taken together, a version of ESLint that supports extension-named processors and an extension-named processor could require me to modify my configuration if I'm specifically targeting code blocks returned by a processor.

Am I lost in the trees here?

nzakas replied:
That is true, but only if the processor is returning filenames with the code blocks. If the processor is still returning strings, then the code blocks get the same name as the parent file. And when ESLint is given string code blocks, it would stop matching against the code block filenames because that would be an infinite loop.

So, it would be a breaking change for the *plugin* if it changes from returning strings to returning filenames, but not a breaking change for ESLint.

btmills replied:
That makes sense. I’m happy to use the Markdown plugin to test the recommended upgrade path for processors once this is available in ESLint.

nzakas commented on designs/2018-processors-improvements/README.md:
```diff
@@ -0,0 +1,284 @@
+
+When a `preprocess()` method returns an object with a `filename` property, a new filename is constructed in this format:
+
+> parentFilename + path.sep + index + filenameFromPreprocess
```
That's correct.

nzakas commented:
@ilyavolodin do you have any further concerns? At this point, you're the only one that has suggested changes outstanding.

nzakas commented:
It looks like we have our first merged RFC! :tada: Thanks everyone!

not-an-aardvark commented:
For future reference, do we need TSC approval to merge RFCs like this? I have no problem with it being merged either way (and I see that a majority of TSC members approved the RFC anyway), but I was a bit confused because I thought it would need to be approved in a TSC meeting.

nzakas commented:
Sorry, I guess that wasn't clear. My thinking was that we don't need a TSC
meeting to approve an RFC when we have enough PR approvals. In this case,
we had five approvals on the PR and no outstanding questions or comments.

If we can't get a consensus on the PR, then I do think we need to discuss
in a TSC meeting.

On Wed, Jan 2, 2019 at 12:13 PM Teddy Katz <notifications@github.com> wrote:

> For future reference, do we need TSC approval to merge RFCs like this? I
> have no problem with it being merged either way (and I see that a majority
> of TSC members approved the RFC anyway), but I was a bit confused because I
> thought it would need to be approved in a TSC meeting.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/rfcs/pull/3#issuecomment-450971824>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACWkmavb7s-GCTa2b8WwCxq5joGk9uIks5u_RL1gaJpZM4Yrmbj>
> .
>


-- 

______________________________
Nicholas C. Zakas
@slicknet

Author, Principles of Object-Oriented JavaScript <http://amzn.to/29Pmfrm>
Author, Understanding ECMAScript 6 <http://amzn.to/29K1mIy>
